### ISSUE TITLE
Where possible, invalid keys should be reset to the default, rather than resetting the whole config.

### ISSUE THREAD

**Issue by cary-rowen** [nvaccess/nvda#14424]
### Is your feature request related to a problem? Please describe.
Related to #14412
While downgrades are not supported, we should take this into account and never assume that users will do certain things exactly as we expect them to.
### Describe the solution you'd like
When upgrading the config, backups should be should kept for each schema version (e.g. just before upgrading, and just after upgrading).
This will allow for better downgrading and upgrading.

### Describe alternatives you've considered
None
### Additional context
None

**Comment by CyrilleB79** [comment]
@cary-rowen the title and the description of this issue seem to describe two different solutions: resetting only invalid keys or keep backups of the config file while upgrading.

I have proposed a related but slightly different solution to this issue in #15779. Let me know if it is enough for you or not.

**Comment by cary-rowen** [comment]
Hi @CyrilleB79 

Indeed, the title and description say two different solutions, but I probably prefer the solution described in the title: only reset invalid keys.
Keeping a copy of a configuration may not make much sense to the Novice user, however, there are many possibilities for a configuration to become corrupted.
In other words, Novice users may cause the configuration to be damaged, but they do not have the ability to fix it manually.
However, I'd still appreciate you trying to come up with #15779, which would be better than the current solution anyway.

Thanks

**Comment by CyrilleB79** [comment]
Before trying to think to a solution, we should understand in which situation such config reinitialization could occur. From what I have in mind, there are the 4 following ones:
1. A user manually edits the nvda.ini file
2. A user downgrades NVDA
3. A user downgrades their configuration using the "-c" flag with various NVDA versions
4. A user reinstalls an older NVDA after having uninstalled a newer NVDA

From these situations, I do not consider 1. and 3. to be representative of a novice user. Situation 2. (downgrading) is not supported by NVDA, and I think that there is at least a warning if you try to do so. And in situation 4., the configuration is trhrown away, but I do not expect the user to expect to keep their previous configuration when completely uninstalling NVDA; so complete reinitialization seems reasonable.

Have you other situations in mind where the config would be corrupted? Especially novice use cases?

**Comment by hwf1324** [comment]
I'm not sure if this is relevant? In the last couple of days, one user moved away from the NVDA installation directory directly, causing NVDA to not start properly.

He didn't downgrade the installation, and it should be reasonable to keep the configuration contents in this case.

**Comment by CyrilleB79** [comment]
> I'm not sure if this is relevant? In the last couple of days, one user moved away from the NVDA installation directory directly, causing NVDA to not start properly.
> 
> He didn't downgrade the installation, and it should be reasonable to keep the configuration contents in this case.

"moved away from the NVDA installation directory directly"?
Sorry, I do not understand this expression. Could you rephrase or clarify with more detailed / explicit steps? Thanks.

**Comment by cary-rowen** [comment]
I think what @hwf1324 means is that the NVDA directory in '%ProgramFiles(x86)%' has been moved to another place. This has absolutely nothing to do with this. We have no reason to consider this extreme situation, not to mention that the user configuration of NVDA is saved in ' In the NVDA directory under %appdata%'.

**Comment by hwf1324** [comment]
Ahh! The reason for this thought is that #15691 fixes this issue.

The reason of course is that I haven't investigated exactly what it did.

**Comment by CyrilleB79** [comment]
For now, the only use cases that can trigger a total reinitialization of the config are the 4 use cases in https://github.com/nvaccess/nvda/issues/14424#issuecomment-1809094404 and the one described in https://github.com/nvaccess/nvda/issues/14424#issuecomment-1809526065.

According to the discussions, among those, none of them is a valid use case to deserve a partial repair of the configuration file (i.e. removing some keys) rather than a total reinitialization of this file.

Thus, I'll close this issue.

If @cary-rowen, @hwf1324 or anyone else can present a valid use case, please describe it here and we can re-open.
